presto-main/src/main/java/com/facebook/presto/operator/project/PageProcessor.java

```diff
@@ -15,11 +15,153 @@
+    {
+        if (page.getPositionCount() == 0) {
+            return EMPTY_PAGE_PROCESSOR_OUTPUT;
+        }
+
+        if (filter.isPresent()) {
+            SelectedPositions selectedPositions = filter.get().filter(session, filter.get().getInputChannels().getInputChannels(page));
+            if (selectedPositions.isEmpty()) {
+                return EMPTY_PAGE_PROCESSOR_OUTPUT;
+            }
+
+            if (projections.isEmpty()) {
+                return new PageProcessorOutput(page.getRetainedSizeInBytes(), Iterators.singletonIterator(new Page(selectedPositions.size())));
+            }
+
+            if (selectedPositions.size() != page.getPositionCount()) {
+                return new PageProcessorOutput(page.getRetainedSizeInBytes(), new PositionsPageProcessorIterator(session, page, selectedPositions));
+            }
+        }
+
+        return new PageProcessorOutput(
+                page.getRetainedSizeInBytes(),
+                new PositionsPageProcessorIterator(session, page, SelectedPositions.positionsRange(0, page.getPositionCount())));
+    }
+
+    private class PositionsPageProcessorIterator
+            extends AbstractIterator<Page>
+    {
+        private final ConnectorSession session;
+        private final Page page;
+
+        private SelectedPositions selectedPositions;
+        private final Block[] previouslyComputedResults;
+
+        public PositionsPageProcessorIterator(ConnectorSession session, Page page, SelectedPositions selectedPositions)
+        {
+            this.session = session;
+            this.page = page;
+            this.selectedPositions = selectedPositions;
+            this.previouslyComputedResults = new Block[projections.size()];
+        }
+
+        @Override
+        protected Page computeNext()
+        {
+            while (true) {
+                if (selectedPositions.isEmpty()) {
+                    return endOfData();
+                }
+
+                int batchSize = Math.min(selectedPositions.size(), projectBatchSize);
+                try {
+                    Page page = processBatch(batchSize);
+
+                    // remove batch from selectedPositions and previouslyComputedResults
+                    selectedPositions = selectedPositions.subRange(batchSize, selectedPositions.size());
+                    for (int i = 0; i < previouslyComputedResults.length; i++) {
+                        if (previouslyComputedResults[i] != null && previouslyComputedResults[i].getPositionCount() > batchSize) {
+                            previouslyComputedResults[i] = previouslyComputedResults[i].getRegion(batchSize, previouslyComputedResults[i].getPositionCount() - batchSize);
+                        }
+                        else {
+                            previouslyComputedResults[i] = null;
+                        }
+                    }
+
+                    return page;
+                }
+                catch (OptimisticProjectException e) {
```

Comment:
Overly code looks good. Some minor concerns:

1. I do not like the idea of using exceptions for controlling flow of execution here.
Maybe consider returning some `ProcessingResult` from `processBatch` which would contain flag that page size limit was exceeded.
Or just return `Optional<Page>` from the method.

2. batch resizing is scattered between `processBatch` and `computeNext`. Would be nice to have it in single place.

3. It would be nice to have in commit message or class comment a description of algorithm for building output pages in given size range. Also a motivation why we do it (what is wrong with too small/too large tables).